presto-main/src/main/java/com/facebook/presto/catalogserver/CatalogServer.java

```diff
@@ -55,103 +71,198 @@ public CatalogServer(MetadataManager metadataProvider, SessionPropertyManager se
     {
         transactionManager.tryRegisterTransaction(transactionInfo);
-        return metadataProvider.catalogExists(session.toSession(sessionPropertyManager), catalogName);
+        CacheKey cacheKey = new CacheKey(session, catalogName);
+        boolean isCacheHit = isCacheHit(cacheKey, catalogExistsCache);
+        Boolean catalogExists = catalogExistsCache.getUnchecked(new CacheKey(session, catalogName));
+        return new CatalogServerClient.MetadataEntry<>(catalogExists, isCacheHit);
     }
 
     @ThriftMethod
-    public String listSchemaNames(TransactionInfo transactionInfo, SessionRepresentation session, String catalogName)
+    public CatalogServerClient.MetadataEntry<String> listSchemaNames(TransactionInfo transactionInfo, SessionRepresentation session, String catalogName)
     {
         transactionManager.tryRegisterTransaction(transactionInfo);
-        List<String> schemaNames = metadataProvider.listSchemaNames(session.toSession(sessionPropertyManager), catalogName);
-        if (!schemaNames.isEmpty()) {
-            return writeValueAsString(schemaNames, objectMapper);
-        }
-        return EMPTY_STRING;
+        CacheKey cacheKey = new CacheKey(session, catalogName);
+        boolean isCacheHit = isCacheHit(cacheKey, listSchemaNamesCache);
+        List<String> schemaNames = listSchemaNamesCache.getUnchecked(cacheKey);
+        return schemaNames.isEmpty()
+                ? new CatalogServerClient.MetadataEntry<>(EMPTY_STRING, isCacheHit)
+                : new CatalogServerClient.MetadataEntry<>(writeValueAsString(schemaNames, objectMapper), isCacheHit);
     }
 
     @ThriftMethod
-    public String getTableHandle(TransactionInfo transactionInfo, SessionRepresentation session, QualifiedObjectName table)
+    public CatalogServerClient.MetadataEntry<String> getTableHandle(TransactionInfo transactionInfo, SessionRepresentation session, QualifiedObjectName table)
     {
         transactionManager.tryRegisterTransaction(transactionInfo);
-        Optional<TableHandle> tableHandle = metadataProvider.getTableHandle(session.toSession(sessionPropertyManager), table);
-        return tableHandle.map(handle -> writeValueAsString(handle, objectMapper))
-                .orElse(EMPTY_STRING);
+        CacheKey cacheKey = new CacheKey(session, table);
+        boolean isCacheHit = isCacheHit(cacheKey, getTableHandleCache);
+        Optional<TableHandle> tableHandle = getTableHandleCache.getUnchecked(cacheKey);
+        if (!tableHandle.isPresent()) {
+            getTableHandleCache.refresh(cacheKey);
+            tableHandle = getTableHandleCache.getUnchecked(cacheKey);
+        }
+        return tableHandle.isPresent()
+                ? new CatalogServerClient.MetadataEntry<>(writeValueAsString(tableHandle.get(), objectMapper), isCacheHit)
+                : new CatalogServerClient.MetadataEntry<>(EMPTY_STRING, isCacheHit);
```

Comment:
`TableHandle.map(...).orElseGet(...)`

Reply:
Done, though i reverted ternary originally because I thought keeping consistent through all methods would be more readable